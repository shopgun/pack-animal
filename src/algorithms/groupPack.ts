import { linePack } from ".";
import packAnimal from "../.";

import { getPolygonTransform, ITransform } from "../transform";

import { IPolygon, polygonArea, polygonHeight, polygonsBounds, polygonWidth } from "../geometry";
import { Matrix } from "../vendor/matrix";

import { average } from "../maths";
import { noop } from "../utilities";

const packRatio = (polygonTransforms: ITransform[]) => {
  const bounds = polygonsBounds(polygonTransforms.map(t => t.points));
  return polygonWidth(bounds) / polygonHeight(bounds);
};

export const groupPack = (
  rectangleWidth: number,
  rectangleHeight: number,
  polygonGroups: IPolygon[][],
  { /* istanbul ignore next */ debug: dbug = noop } = {}
): ITransform[] => {
  const rectangleRatio = rectangleWidth / rectangleHeight;
  // Wrap debug function to include current algorithm.
  const debug = (...args: any[]) => dbug("groupPack:", ...args);
  // Write out said algorithm entry.
  debug();
  const averageArea = average(
    Array.prototype.concat(...polygonGroups).map(({ points }) => polygonArea(points))
  );

  const packs = polygonGroups.map(polygonGroup => ({
    polygons: polygonGroup,
    transforms: packAnimal(
      rectangleHeight,
      rectangleWidth,
      polygonGroup.map(({ points }) => points),
      {
        averageArea,
        center: true,
        debug: dbug !== noop,
        maximize: false
      }
    )
  }));
  const packsBounds = packs.map(polygonTransforms =>
    polygonsBounds(
      polygonTransforms.transforms.map((polygonTransform: ITransform) => polygonTransform.points)
    )
  );
  const packedPacks = [
    linePack(
      rectangleWidth,
      rectangleHeight,
      packsBounds.map((points, index) => ({ points, index })),
      false,
      {
        alignBottom: false,
        averageArea: average(packsBounds.map(packBounds => polygonArea(packBounds))),
        debug
      }
    ),
    linePack(
      rectangleWidth,
      rectangleHeight,
      packsBounds.map((points, index) => ({ points, index })),
      true,
      { debug, averageArea: average(packsBounds.map(packBounds => polygonArea(packBounds))) }
    )
  ].sort(
    (a, b) => Math.abs(packRatio(a) - rectangleRatio) - Math.abs(packRatio(b) - rectangleRatio)
  )[0];

  const transforms = packedPacks.reduce((memo: ITransform[], packTransform, index) => {
    const pack = packs[index];
    pack.polygons.forEach((polygon, polygonGroupPolygonIndex) => {
      const polyTransform = pack.transforms[polygonGroupPolygonIndex];
      const points = polyTransform.matrix.inverse().applyToArray(polyTransform.points);
      const sumTransform = getPolygonTransform(
        rectangleWidth,
        rectangleHeight,
        points,
        Matrix.from(
          polyTransform.matrix.a,
          polyTransform.matrix.b,
          polyTransform.matrix.c,
          polyTransform.matrix.d,
          polyTransform.matrix.e + packTransform.matrix.e,
          polyTransform.matrix.f + packTransform.matrix.f
        )
      );

      memo[polygon.index] = sumTransform;
    });
    return memo;
  }, []);
  return transforms;
};
