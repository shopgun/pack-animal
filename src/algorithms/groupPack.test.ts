import { groupPack } from "./groupPack";

describe("groupPack", () => {
  it("works", () => {
    expect(
      groupPack(
        800,
        742,
        [
          [
            [{ x: 0, y: 0 }, { x: 2, y: 0 }, { x: 2, y: 8 }, { x: 0, y: 8 }],
            [{ x: 0, y: 0 }, { x: 1, y: 0 }, { x: 1, y: 12 }, { x: 0, y: 12 }]
          ],
          [
            [{ x: 0, y: 0 }, { x: 12, y: 0 }, { x: 12, y: 1 }, { x: 0, y: 1 }],
            [{ x: 0, y: 0 }, { x: 8, y: 0 }, { x: 8, y: 2 }, { x: 0, y: 2 }]
          ],
          [
            [{ x: 0, y: 0 }, { x: 4, y: 0 }, { x: 4, y: 4 }, { x: 0, y: 4 }],
            [{ x: 0, y: 0 }, { x: 8, y: 0 }, { x: 8, y: 8 }, { x: 0, y: 8 }]
          ]
        ].map((group, groupIndex) =>
          group.map((points, index) => ({
            index: (index + 1) * (groupIndex + 1) - 1,
            points
          }))
        )
      )
    ).toMatchSnapshot();
  });
});
