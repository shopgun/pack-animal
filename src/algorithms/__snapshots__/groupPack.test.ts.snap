// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groupPack works 1`] = `
Array [
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 0.3020833333333428%;
        height: 1.3027852650494263%;
        z-index: 414;
        transform: translateX(15271.839080459293%) translateY(4243.021886124295%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":1.2083333333333335,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":1.2083333333333335,
      \\"e\\":369.06944444444446,
      \\"f\\":410.15878232535175
    }",
    "points": Array [
      Object {
        "x": 369.06944444444446,
        "y": 410.15878232535175,
      },
      Object {
        "x": 371.4861111111112,
        "y": 410.15878232535175,
      },
      Object {
        "x": 371.4861111111112,
        "y": 419.8254489920185,
      },
      Object {
        "x": 369.06944444444446,
        "y": 419.8254489920185,
      },
    ],
    "rotate": -0,
    "scale": 1.2083333333333335,
    "svgTransform": "matrix(1.2083333333333335,0,0,1.2083333333333335,369.06944444444446,410.15878232535175)",
    "translateX": 369.06944444444446,
    "translateY": 410.15878232535175,
    "zIndex": 414,
  },
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 2.1694878017789208%;
        height: 0.23540873676624732%;
        z-index: 422;
        transform: translateX(2077.3873343669993%) translateY(24115.810443074733%) rotate(1.0000000000002647deg)
      ",
    "matrix": "{
      \\"a\\":1.444224448559232,
      \\"b\\":0.02520903152052063,
      \\"c\\":-0.02520903152052063,
      \\"d\\":1.444224448559232,
      \\"e\\":360.5493185183387,
      \\"f\\":421.2387774593886
    }",
    "points": Array [
      Object {
        "x": 360.5493185183386,
        "y": 421.23877745938853,
      },
      Object {
        "x": 377.88001190104944,
        "y": 421.5412858376347,
      },
      Object {
        "x": 377.85480286952884,
        "y": 422.9855102861941,
      },
      Object {
        "x": 360.5241094868181,
        "y": 422.68300190794764,
      },
    ],
    "rotate": 1.0000000000002647,
    "scale": 1.4444444444444446,
    "svgTransform": "matrix(1.444224448559232,0.02520903152052063,-0.02520903152052063,1.444224448559232,360.5493185183387,421.2387774593886)",
    "translateX": 360.5493185183387,
    "translateY": 421.2387774593886,
    "zIndex": 422,
  },
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 0.6041666666666643%;
        height: 0.6513926325246977%;
        z-index: 418;
        transform: translateX(8258.59596088066%) translateY(8608.51546746951%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":1.2083333333333335,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":1.2083333333333335,
      \\"e\\":399.16547144256367,
      \\"f\\":416.07824759435306
    }",
    "points": Array [
      Object {
        "x": 399.16547144256367,
        "y": 416.0782475943531,
      },
      Object {
        "x": 403.998804775897,
        "y": 416.0782475943531,
      },
      Object {
        "x": 403.998804775897,
        "y": 420.9115809276864,
      },
      Object {
        "x": 399.16547144256367,
        "y": 420.9115809276864,
      },
    ],
    "rotate": -0,
    "scale": 1.2083333333333335,
    "svgTransform": "matrix(1.2083333333333335,0,0,1.2083333333333335,399.16547144256367,416.07824759435306)",
    "translateX": 399.16547144256367,
    "translateY": 416.07824759435306,
    "zIndex": 418,
  },
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 1.2083333333333357%;
        height: 0.32569631626235657%;
        z-index: 418;
        transform: translateX(3949.158465734759%) translateY(17251.101837260936%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":1.2083333333333335,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":1.2083333333333335,
      \\"e\\":381.75198502102745,
      \\"f\\":416.9016277338092
    }",
    "points": Array [
      Object {
        "x": 381.7519850210275,
        "y": 416.9016277338092,
      },
      Object {
        "x": 391.4186516876942,
        "y": 416.9016277338092,
      },
      Object {
        "x": 391.4186516876942,
        "y": 419.3182944004759,
      },
      Object {
        "x": 381.7519850210275,
        "y": 419.3182944004759,
      },
    ],
    "rotate": -0,
    "scale": 1.2083333333333335,
    "svgTransform": "matrix(1.2083333333333335,0,0,1.2083333333333335,381.75198502102745,416.9016277338092)",
    "translateX": 381.75198502102745,
    "translateY": 416.9016277338092,
    "zIndex": 418,
  },
  undefined,
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 0.6770833333333357%;
        height: 0.7300089847259761%;
        z-index: 415;
        transform: translateX(7313.8240881703805%) translateY(7626.05995558795%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":0.6770833333333334,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":0.6770833333333334,
      \\"e\\":396.16547144256367,
      \\"f\\":413.07824759435306
    }",
    "points": Array [
      Object {
        "x": 396.16547144256367,
        "y": 413.07824759435306,
      },
      Object {
        "x": 401.58213810923036,
        "y": 413.07824759435306,
      },
      Object {
        "x": 401.58213810923036,
        "y": 418.4949142610198,
      },
      Object {
        "x": 396.16547144256367,
        "y": 418.4949142610198,
      },
    ],
    "rotate": -0,
    "scale": 0.6770833333333334,
    "svgTransform": "matrix(0.6770833333333334,0,0,0.6770833333333334,396.16547144256367,413.07824759435306)",
    "translateX": 396.16547144256367,
    "translateY": 413.07824759435306,
    "zIndex": 415,
  },
]
`;
